File: Data-Structures/Trie.py
"""
- Trie -

A tree with as many pointers from every node as the number of
characters in the alphabet.

Applications:
- Auto complete
"""

class Node:
    def __init__(self, char):
        self.char = char
        self.children = {}
        self.end_of_word = False


class Trie:
    def __init__(self):
        self.root = Node('*')
    
    def insert(self, word):
        current = self.root
        for char in word:
            if char in current.children:
                current = current.children[char]
            else:
                new_node = Node(char)
                current.children[char] = new_node
                current = new_node
        current.end_of_word = True
    
    def search(self, word):
        if self.root.children is None:
            return False
        current = self.root
        for char in word:
            if char in current.children:
                current = current.children[char]
            else:
                return False
        if current.end_of_word:
            return True
        return False

trie = Trie()
trie.insert('hello')
trie.insert('goodbye')
print(trie.search('hello'))
print(trie.search('hell'))
print(trie.search('goodbye'))
print(trie.search('good'))

File: DataStructures/Trees/Trie.js
/*
- Trie -

A tree with as many pointers from every node as the number of
characters in the alphabet.

Applications:
- Auto complete
*/

class Node {
    constructor(char) {
        this.char = char
        this.children = {}
        this.endOfWord = false
    }
}

class Trie {
    constructor() {
        this.root = new Node('*')
    }

    insert(word) {
        let current = this.root
        for (let char of word) {
            if (current.children.hasOwnProperty(char)) {
                current = current.children[char]
            } else {
                let newNode = new Node(char)
                current.children[char] = newNode
                current = newNode
            }
        }
        current.endOfWord = true
    }

    search(word) {
        if (this.root.children === null) {
            return false
        }
        let current = this.root
        for (let char of word) {
            if (current.children.hasOwnProperty(char)) {
                current = current.children[char]
            } else {
                return false
            }
        }
        if (current.endOfWord) {
            return true
        } else {
            return false
        }
    }
}

const trie = new Trie()
trie.insert('hello')
trie.insert('goodbye')
console.log(trie.search('hello'))
console.log(trie.search('hell'))
console.log(trie.search('goodbye'))
console.log(trie.search('good'))
